extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/runtime/util/VersionHandler.java

```diff
@@ -0,0 +1,146 @@
+package io.quarkus.mongodb.panache.common.runtime.util;
+
+import java.lang.reflect.Field;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.bson.BsonDocument;
+import org.bson.BsonNull;
+import org.bson.BsonValue;
+
+import io.quarkus.mongodb.panache.common.Version;
+import io.quarkus.mongodb.panache.common.exception.OptimisticLockException;
+import io.quarkus.mongodb.panache.common.runtime.MongoOperations;
+
+/**
+ * Handle @Version in queries.
+ **/
+public class VersionHandler {
+
+    public final Object entity;
+    public final Field versionField;
+    public final BsonDocument query;
+    public final Boolean containsVersionAnnotation;
+    public final Boolean containsVersionValue;
+    public final String versionFieldName;
+
+    private VersionHandler(final Object entity,
+            final BsonDocument query,
+            Field versionField,
+            Boolean containsVersionAnnotation,
+            Boolean containsVersionValue,
+            String versionFieldName) {
+        this.entity = entity;
+        this.query = query;
+        this.versionField = versionField;
+        this.containsVersionAnnotation = containsVersionAnnotation;
+        this.containsVersionValue = containsVersionValue;
+        this.versionFieldName = versionFieldName;
+
+        if (versionField != null) {
+            adjustVersionValue();
+        }
+    }
+
+    public static VersionHandler handle(Object entity, BsonDocument document) {
+        Optional<Field> versionFieldOptional = extractVersionField(entity);
+
+        if (versionFieldOptional.isEmpty()) {
+            return new VersionHandler(
+                    entity,
+                    buildUpdateQuery(null, document),
+                    null,
+                    false,
+                    false,
+                    null);
+        }
+
+        Field versionField = versionFieldOptional.get();
+        Long versionValue = extractVersionValue(versionField, entity);
+        String versionFieldName = versionField.getName();
+        BsonDocument query = buildUpdateQuery(versionFieldName, document);
+
+        return new VersionHandler(entity, query, versionField, true, versionValue != null, versionFieldName);
+    }
+
+    public boolean containsVersionAnnotationAndValue() {
+        return containsVersionAnnotation && containsVersionValue;
+    }
+
+    private static Optional<Field> extractVersionField(Object entity) {
```

Comment:
It would be better to use Jandex at build time to compute a map of version field by class, this would avoid using reflection at runtime which we want to avoid.
I also don't like the fact that if multiple fields exist in the class hierarchy you just take the first one, you should either use the closest one on the hierarchy or generate an error.

Reply:
I think we should raise an error, just adjusted here.

Reply:
@loicmathieu do you remember any place in panache making use of this? I can change here, btw I'd like to ask you for review again, and if all is good, can I change to jandex in the next PR?